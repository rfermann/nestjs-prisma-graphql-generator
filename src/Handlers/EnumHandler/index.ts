import type { Enum } from "../../types";
import { NestJSTypes } from "../../types";
import { BaseHandler } from "../BaseHandler";
import { compareObjectValues } from "../BaseHandler/compareFunctions";

export class EnumHandler extends BaseHandler {
  private readonly _enums: Enum[] = [];

  async createBarrelFile(): Promise<void> {
    if (this._enums.length) {
      const sourceFile = this.baseFileGenerator.createSourceFile(`${this.config.paths.enum}/index`);

      sourceFile.addExportDeclarations(
        this._enums.map(({ name }) => {
          return {
            moduleSpecifier: `./${name}`,
            namedExports: [name],
          };
        })
      );
      await sourceFile.save();
    }
  }

  async createFiles(): Promise<void> {
    // eslint-disable-next-line no-restricted-syntax
    for (const { documentation, name: enumName, values } of this._enums) {
      const sourceFile = this.baseFileGenerator.createSourceFile(`${this.config.paths.enum}/${enumName}`);

      this.baseFileGenerator.addNestJSImports({ nestJSImports: [NestJSTypes.RegisterEnumType], sourceFile });

      sourceFile.addEnum({
        isExported: true,
        members: values.map((data) => {
          return {
            name: data.name,
            value: data.value,
          };
        }),
        name: enumName,
      });

      sourceFile.addStatements((writer) =>
        writer
          .writeLine("")
          .writeLine(`${NestJSTypes.RegisterEnumType}(${enumName}, {`)
          .writeLine(`name: "${enumName}",`)
          .conditionalWrite(typeof documentation === "string", `description: "${documentation}"`)
          .writeLine("});")
      );

      sourceFile.formatText();
    }

    await this.baseFileGenerator.save();
  }

  parse(): void {
    this.baseParser.dmmf.datamodel.enums.forEach(({ documentation, name, values }): void => {
      this._enums.push({
        documentation,
        name: this.baseParser.getEnumName(name),
        values: values.map(({ name: value }) => {
          return { name: value, value };
        }),
      });
    });

    this.baseParser.dmmf.schema.enumTypes.prisma.forEach(({ name, values }): void => {
      this._enums.push({
        name: this.baseParser.getEnumName(name),
        values: values.map((value) => {
          return { name: value, value };
        }),
      });
    });

    this._enums.sort((a, b) => compareObjectValues({ a, b, field: "name" }));
  }
}
