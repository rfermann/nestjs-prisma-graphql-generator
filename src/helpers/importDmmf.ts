/* eslint-disable global-require */
/* eslint-disable import/no-dynamic-require */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import type { DMMF } from "@prisma/generator-helper";

export const importDmmf = (prismaClientImportPath: string): DMMF.Document =>
  require(prismaClientImportPath).dmmf as DMMF.Document;
