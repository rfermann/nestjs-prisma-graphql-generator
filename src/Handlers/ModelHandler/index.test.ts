// eslint-disable-next-line import/no-nodejs-modules
import { existsSync, readFileSync, rmdirSync } from "fs";

import type { GeneratorConfig } from "../../GeneratorConfig";
import { importDmmf } from "../../helpers";
import { EnumHandler } from "../EnumHandler";

import { ModelHandler } from ".";

describe("ModelHandler", () => {
  // eslint-disable-next-line jest/no-hooks
  beforeAll(async () => {
    if (existsSync(`${process.cwd()}/fixtures/modelHandler/fixtures/`)) {
      rmdirSync(`${process.cwd()}/fixtures/modelHandler/fixtures/`, { recursive: true });
    }
  });
  it("should parse DMMF properly", async () => {
    expect.assertions(1);

    expect(true).toBe(true);
  });

  it("should parse models and create correct files from datamodel", async () => {
    expect.assertions(2);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/modelHandler/fixtures/output3`,
      paths: {
        enums: "enums",
        inputTypes: "",
        model: "model",
        shared: "",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
    };

    const enumHandler = new EnumHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    const modelHandler = new ModelHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    enumHandler.parse();

    modelHandler.parse(enumHandler.getEnums());
    await modelHandler.createFiles();

    const userModelFile = readFileSync(`${config.basePath}/User/model.ts`, "utf-8");
    const sessionModelFile = readFileSync(`${config.basePath}/Session/model.ts`, "utf-8");

    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(userModelFile).toMatchSnapshot();

    expect(sessionModelFile).toMatchInlineSnapshot(`
      "import { Field, GraphQLISODateTime, Int, ObjectType } from \\"@nestjs/graphql\\";
      import { ByteResolver } from \\"graphql-scalars\\";
      import { User } from \\"../User/model\\";

      @ObjectType({
        isAbstract: true,
      })
      export class Session {
        @Field(() => Int, {
          nullable: false
        })
        id!: number;

        @Field(() => GraphQLISODateTime, {
          nullable: false
        })
        createdAt!: Date;

        @Field(() => User, {
          nullable: true
        })
        user?: User | null;

        @Field(() => Int, {
          nullable: false
        })
        userId!: number;

        @Field(() => Decimal, {
          nullable: false
        })
        size!: number;

        @Field(() => BigInt, {
          nullable: false
        })
        bigSize!: bigint;

        @Field(() => ByteResolver, {
          nullable: false
        })
        byte!: Buffer;
      }
      "
    `);
  });
});
