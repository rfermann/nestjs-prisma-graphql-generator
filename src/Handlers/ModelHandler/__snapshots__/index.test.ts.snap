// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelHandler should parse models and create correct files from datamodel 1`] = `
"import { Field, Float, Int, ObjectType } from \\"@nestjs/graphql\\";
import { GraphQLJSON } from \\"graphql-type-json\\";
import { JsonValue } from \\"@prisma/client\\";
import { UserTypeEnum } from \\"../enums\\";
import { Session } from \\"../Session/model\\";

/** user model comment */
@ObjectType({
  isAbstract: true,
  description: \\"user model comment\\"
})
export class User {
  @Field(() => Int, {
    nullable: false
  })
  id!: number;

  @Field(() => String, {
    nullable: false
  })
  email!: string;

  @Field(() => String, {
    nullable: true
  })
  name?: string | null;

  @Field(() => UserTypeEnum, {
    nullable: false
  })
  type!: \\"USER\\" | \\"ADMIN\\";

  @Field(() => Boolean, {
    nullable: false
  })
  active!: boolean;

  @Field(() => Float, {
    nullable: false
  })
  balance!: number;

  /** user field comment */
  @Field(() => GraphQLJSON, {
    nullable: false
    description: \\"user field comment\\"
  })
  rawData!: JsonValue;

  @Field(() => [Session], {
    nullable: true
  })
  Session?: Session[];
}
"
`;
